### PR TITLE
POC: Tests for `configure-ovs.sh`

### PR DESCRIPTION
**Pull request by zeeke**
Testing `configure-ovs.sh`.sh script needs a dedicated Virtual Machine, as mocking its dependencies would lead to a too simplistic test case.

Tests leverage `bats` as bash test framework and [kcli](https://github.com/karmab/kcli) to spin up VMs with different network configurations.

As a POC, two tests have been implemented:
- `Single NIC` is about a VM with a plain ethernet NIC
- `Bonding NICs` checks creating br-ex on top of a bond interface.

Each VM is configured to configure its networking with startup commands and then run `configure-ovs.sh` (see plan/*.yml files)

Test cases make assertions on the stdout and stderr of the script, brought out from the VMs by `finishfiles` fields.

Thoughts?

### CODE REVIEW

File: test/kcli/plans/bonding-nics.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+vm3:
+  image: fedora37
+  numcpus: 2
+  memory: 2048
+  # output logs from following commands can be found in guest path /var/log/cloud-init-output.log
+  cmds:
+  - yum -y install openvswitch NetworkManager-ovs jq nmstate
+  - systemctl start openvswitch
+  - nmcli connection down "cloud-init eth1"
```
Comment (cybertron) [inline test/kcli/plans/bonding-nics.yml:13]: Since we're installing nmstate anyway, we could use it to configure the bond. That would simplify this test case quite a bit.

**Review by cybertron — COMMENTED** [review]
Looks useful to me. Is there any way to do validation of the results of the script or would we need a wrapper around this to look at the nmstate.txt output?

File: test/kcli/plans/bonding-nics.yml (hunk 1 above)
Comment (zeeke) [reply test/kcli/plans/bonding-nics.yml:13]: Great point! I'm not an expert on nmstate, do you know if it can also delete automatic connections? It would be great if you could provide some snippets to start from.

**Comment by zeeke** [comment]
> Looks useful to me. Is there any way to do validation of the results of the script or would we need a wrapper around this to look at the nmstate.txt output?

For this POC, I only put asserts on script stdout+stderr. VMs are also [configured to dump nmstate](https://github.com/openshift/machine-config-operator/pull/3527/files#diff-e047234a46a389d622a3c3b05ee41d8db4ee38ea77c364d483af8996bdbb7e96R15) after calling configure-ovs.sh, so the test can also make assertion on that, maybe using tools like `yq`.

File: test/kcli/configure-ovs-network-test.bats [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    cleanUpCmd=""
+}
+
+teardown() {
+    run ${cleanUpCmd}
+}
+
+@test "Single NIC" {
+    output_dir=${DIR}/_output/single-nic
+    rm -rf ${output_dir}
```
Comment (rbbratta) [inline test/kcli/configure-ovs-network-test.bats:23]: always quote?  `"${output_dir}"` and maybe `--` guard for extra paranoia.  `rm -rf -- `

File: test/kcli/plans/bond99.nmconnection [hunk 3]
```diff
@@ -0,0 +1,25 @@
+[connection]
+id=bond99
```
Comment (rbbratta) [inline test/kcli/plans/bond99.nmconnection:2]: We had `configure-ovs.sh` bugs with spaces in filenames for bonding slaves.

```
id=System eth1
id=Wired connection 1
id=Wired connection enp6s0
id=Wired connection enp5s0-slave-ovs-clone
```
are standard auto-generated names that cause spaces in the `.nmconnection` files.

See https://bugzilla.redhat.com/show_bug.cgi?id=2101157
and
https://cgit.freedesktop.org/NetworkManager/NetworkManager/tree/src/core/settings/plugins/ifcfg-rh/nms-ifcfg-rh-reader.c#n373

I guess maybe we handle this below with `"cloud-init eth1"`

File: test/kcli/configure-ovs-network-test.bats (hunk 2 above)
Comment (zeeke) [reply test/kcli/configure-ovs-network-test.bats:23]: Good point! fixing

File: test/kcli/plans/bond99.nmconnection (hunk 3 above)
Comment (zeeke) [reply test/kcli/plans/bond99.nmconnection:2]: Adding a space to the bond connection name. We can add tests with specific naming later on

**Comment by jcaamano** [comment]
@zeeke awesome

Should this live in a more configure-ovs specific path? Such as `test/configure-ovs`? @yuqi-zhang can you provide some feedback on where this might live?

Also, would it be useful to match (diff) the test output keyfiles (nmconnection files) against expected keyfiles and only pass the test if they are the same (ignoring thee UUID and maybe some other stuff I guess)?

**Comment by zeeke** [comment]
@jcaamano thanks for the feedback

> Should this live in a more configure-ovs specific path? Such as test/configure-ovs? 
I initially thought to put other script tests in this folder, so a generic name should work. But I'm open to suggestions, as I don't like `test/kcli` so much.

> Also, would it be useful to match (diff) the test output keyfiles (nmconnection files) against expected keyfiles and only pass the test if they are the same (ignoring thee UUID and maybe some other stuff I guess)?

I don't have strong opinions on how to assert the final state. Diffing output files and/or using nmstate dump (as suggested by @cybertron) are good alternatives and we can elaborate more on them in further PRs.

File: test/kcli/plans/bonding-nics.yml (hunk 1 above)
Comment (cybertron) [reply test/kcli/plans/bonding-nics.yml:13]: NMState is declarative, so if you tell it to bond two connections it should do whatever is necessary to make that happen. The reason we have trouble with the automatic connections in some envs is that we don't actually run NMState, we just lay down nmconnection files generated by NMState.

The NMState docs have an example of creating a bond (and a lot of other things): https://nmstate.io/examples.html#interfaces-bond

**Comment by yuqi-zhang** [comment]
> Should this live in a more configure-ovs specific path? Such as test/configure-ovs?

I think that could work, so long as it is self-contained. Do you plan on configuring this to run as a separate (potentially optional) test within the MCO test suite? Or do you want to consider this a core test to integrate within an existing test?

File: test/kcli/configure-ovs-network-test.bats [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+   
+    cleanUpCmd=""
+}
+
+teardown() {
+    run ${cleanUpCmd}
+}
+
+@test "Single NIC" {
+    output_dir=${DIR}/_output/single-nic
```
Comment (martinkennelly) [inline test/kcli/configure-ovs-network-test.bats:19]:
```suggestion
    output_dir="${DIR}/_output/single-nic"
```

File: test/kcli/configure-ovs-network-test.bats [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    rm -rf -- "${output_dir}"
+    mkdir -p "${output_dir}"
+
+    run kcli create plan -f plans/single-nic.yml -P output_dir=${output_dir}
+    cleanUpCmd="kcli delete -y vm vm3"
+
+    output_file=${output_dir}/configure-ovs-output.txt
```
Comment (martinkennelly) [inline test/kcli/configure-ovs-network-test.bats:26]:
```suggestion
    output_file="${output_dir}/configure-ovs-output.txt"
```

File: test/kcli/configure-ovs-network-test.bats [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    assert_file_contains "${output_file}" "Brought up connection br-ex successfully"
+    assert_file_contains "${output_file}" "Brought up connection ovs-if-br-ex successfully"
+}
+
+@test "Bonding NICs" {
+    output_dir=${DIR}/_output/bonding-nics/
```
Comment (martinkennelly) [inline test/kcli/configure-ovs-network-test.bats:32]:
```suggestion
    output_dir="${DIR}/_output/bonding-nics/"
```

File: test/kcli/configure-ovs-network-test.bats [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    rm -rf -- "${output_dir}"
+    mkdir -p "${output_dir}"
+
+    run kcli create plan -f plans/bonding-nics.yml -P output_dir=${output_dir}
```
Comment (martinkennelly) [inline test/kcli/configure-ovs-network-test.bats:36]:
```suggestion
    run kcli create plan -f plans/bonding-nics.yml -P output_dir="${output_dir}"
```

File: test/kcli/configure-ovs-network-test.bats [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    cleanUpCmd="kcli delete -y vm vm3"
+
+    output_file=${output_dir}/configure-ovs-output.txt
```
Comment (martinkennelly) [inline test/kcli/configure-ovs-network-test.bats:39]:
```suggestion
    output_file="${output_dir}/configure-ovs-output.txt"
```

File: test/kcli/plans/bond99.nmconnection [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+[ipv4]
+method=auto
+may-fail=false
+dhcp-timeout=2147483647
+route-metric=50
+
+[ipv6]
+method=disabled
+may-fail=true
+ra-timeout=2147483647
```
Comment (martinkennelly) [inline test/kcli/plans/bond99.nmconnection:25]: nit: missing \n

**Review by martinkennelly — COMMENTED** [review]
Great work. Just nits really. Unfamiliar with the bats framework.
Looking forward to seeing nmstate do the bonds like what ben said.

File: test/kcli/plans/single-nic.yml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+parameters:
+  output_dir: ../_output/test
+
+vm3:
+  image: fedora37
+  numcpus: 2
+  memory: 2048
+  # output logs from following commands can be found in guest path /var/log/cloud-init-output.log
+  cmds:
+  - yum -y install openvswitch NetworkManager-ovs jq nmstate
```
Comment (martinkennelly) [inline test/kcli/plans/single-nic.yml:11]: super nit: dnf is recommended instead of yum.

**Comment by zeeke** [comment]
@jcaamano @cybertron 

I improved the assertion mechanism using yq. Tests are now a little bit less readable, but thanks to those assertions, I found an edge case to fix in configure-ovs.sh script:

- https://github.com/openshift/machine-config-operator/pull/3548

So, I'd they it's worth the complexity

**Comment by yuqi-zhang** [comment]
Just to check in, would we still like to add this to the MCO test suite? Did you want to e.g. make it required and run as part of MCO's e2e? Should this be part of an optional job? Should this be an independent test?

**Comment by zeeke** [comment]
Hi @yuqi-zhang , thanks for bringing this up.

I used this test suite on my local machine to troubleshoot some configure-ovs.sh issue. Hence I think it has some value.
I'm still trying to refactor test cases to use nmstate for the initial configuration.

After that, I'd like to put it in an optional CI job, but it requires libvirt to be installed. I don't know if there are some CI step/workflow that we can use for that.

**Comment by yuqi-zhang** [comment]
Thanks for the update, I'm not really sure if we have some type of libvirt profile we can leverage to test this on. Maybe a better question for the testplatform team.

An optional CI job that is separate probably makes the most sense to me as well
